traits/tests/test_regression.py

```diff
@@ -289,3 +299,34 @@ class A(HasTraits):
 
         with self.assertRaises(ZeroDivisionError):
             a.bar = "foo"
+
+
+class NestedContainerClass(HasTraits):
+    # Used in regression test for changes to nested containers
+    # e.g. enthought/traits#281
+    dict_of_list = Dict(Str, List(Str))
+
+    dict_of_union_none_or_list = Dict(Str, Union(List(), None))
+
+
+class TestRegressionNestedContainerEvent(unittest.TestCase):
+    """ Regression tests for enthought/traits#281
+    """
+
+    def test_modify_list_in_dict(self):
```

Comment:
Indeed now that #913 has gone in, this test no longer fails when the addition to `TraitListObject` was removed. Thanks for testing that.

While I agree with you that lines 2525-2529 look like it is trying to handle the same situation, I am not confident that we can remove it just yet.

The `self.value_trait.handler` is an instance of `TraitHandler`.  There are a few instances of `TraitHandler` that have `has_items` set to `True` (not exhaustive): `TraitList(TraitHandler)`, `TraitDict(TraitHandler)`, `TraitSet(TraitHandler)`. They are deprecated on master, but the deprecation has not yet been released.  So it seems a bit early to remove this block.

FTR, while `TraitCompound` (used by `Either`) is also a subclass of `TraitHandler` that could have `has_items` set to True, the usage of `Dict(Any(), Either(List, None))` never worked, because the `clone` method is not implemented on `TraitCompound`.

At the time when `TraitList(TraitHandler)`, `TraitDict(TraitHandler)`, `TraitSet(TraitHandler)` are removed, we could revisit the possibility of removing this block (repeating for clarification):
https://github.com/enthought/traits/blob/0f20d41fd0ceba6205e90ed9960bbcf281ab965e/traits/trait_types.py#L2525-L2529
What do you think?

Just to confirm, this test is still being kept as a regression test for #281, despite it being resolved by #931 (it looks somewhat accidental to me to be honest).

Reply:
I didn't see the deprecated `TraitHandler`s. In that case I agree that it's too early to remove the block.

And keeping the test sounds good to me!

Reply:
:+1:

Correction (probably won't affect the conclusion): The statement "They are deprecated on master, but the deprecation has not yet been released. " was incorrect. They are released in 6.0.0. Sorry about that.

Reply:
Opened #1052 to remind ourselves to consider this `has_items` usage again.